File: C++/handleData.cpp
#include "Crime.h"
#include <vector>
#include <fstream>
#include <sstream>

using namespace std;

// Stores the indices of the different imported field names present in the data
enum fields
{
    incident = 1,
    date = 3,
    address = 10,
    latitude = 11,
    longitude = 12
};

// Splits line at commas and places data into a vector
vector<string> split(string str)
{
    stringstream stream(str);
    vector<string> data;

    string value = "";
    while (getline(stream, value, ','))
    {
        data.push_back(value);
    }

    return data;
}

vector<Crime> getData(string filepath)
{
    // Stores all crimes with valid data
    vector<Crime> crimes = {};

    // Opens file at inputted pathat and stores it in file variable
    ifstream file(filepath);

    // Stores the current line in CSV file
    string line = "";

    // Ignore first line since it is the unformatted field names
    getline(file, line);

    // Loops through all data in CSV files
    while (getline(file, line))
    {
        // Splits current line at delimiter of ','
        vector<string> data = split(line);

        // Only creates crime objects out of crimes with valid data
        try
        {
            Crime newCrime(data[incident], data[date], data[address], stof(data[longitude]), stof(data[latitude]));
            crimes.push_back(newCrime);
        }
        catch (exception _)
        {
            continue;
        }
    }

    return crimes;
}

int main()
{
    vector<Crime> crimes = getData("Crime_Responses.csv");
    crimes[0].display();
    return 0;
}

File: JavaScript/handleData.js
// 
import { Crime } from './Crime.js';

// 
const MAX_RANGE = 100;

// 
var crimes = null;

// 
var inputForm = document.getElementById('input-form');
var indexInput = document.getElementById("index-input");
var crimeText = document.getElementById("crime-text")

// 
function parse(data) {
    const lines = data.split("\n");

    // Remove first line since it just contains columns names
    lines.splice(0, 1);

    var crimes = [];

    // 
    lines.forEach((line) => {
        var data = line.split(",");

        // Only adds crime if latitude and longitude data is provided
        if (parseFloat(data[11]) && parseFloat(data[12])) {
            var crime = new Crime(data[1], data[3], data[10], data[11], data[12]);
            crime.setDistance(29.69052, -82.33503);

            // 
            if (crime.distance <= MAX_RANGE) crimes.push(crime);
        }
    });

    return crimes;
}

//
async function getCrimeData(filepath) {
    // 
    return fetch(filepath)
        .then(file => file.text())
        .then(data => parse(data));
}

// 
inputForm.addEventListener('submit', function (event) {
    // 
    event.preventDefault();

    // 
    var index = (indexInput.value == "") ? 1 : indexInput.value;

    // 
    crimeText.innerHTML = crimes[index - 1].getData();
});


// 
async function main() {
    // 
    crimes = await getCrimeData("../Crime_Responses.csv");

    // 
    crimeText.textContent = `${crimes.length} crimes loaded!`;
    indexInput.max = crimes.length;
}

main()